content/blog/neutron-where-is-my-packet-2.md

```diff
@@ -0,0 +1,1082 @@
+ cookie=0xc609e5deded3773a, duration=7102.714s, table=62, n_packets=0, n_bytes=0, priority=3 actions=NORMAL
+```
+
+The most important and interesting rule here is rule:
+
+```
+ cookie=0xc609e5deded3773a, duration=5754.608s, table=0, n_packets=767, n_bytes=116954, priority=3,in_port="int-br-ex",vlan_tci=0x0000/0x1fff actions=mod_vlan_vid:2,resubmit(,60)
+```
+
+It matches packets which are comming from the ``int-br-ex`` patch port and have
+no vlan_id (``vlan_tci=0x0000/0x1fff``). For those packets it sets vlan_id=2
+(it's again the vlan_id used locally by the neutron-ovs-agent and can be
+different on each node for the same network, or can even be changed after
+restart of e.g. openvswitch). Then it is send to the table 60 where it hits rule
+with action ``NORMAL`` again so it is send to the proper interface, which in our
+case will be the ``qg-`` port.
+
+
+#### Debugging connectivity issues through the SNAT
+
+For connectivity using SNAT most of the parts are exactly the same as when
+Floating IP is used. The only differences here are in the iptables rules in the
+``qrouter`` namespace:
+
+```bash
+$ sudo ip netns exec qrouter-50711ca1-1e58-44dc-bd95-a87250d5b4ae iptables-save
+...
+# Generated by iptables-save v1.8.4 on Mon Jan 24 14:28:27 2022
+*nat
+...
+-A neutron-l3-agent-snat -j neutron-l3-agent-float-snat
+-A neutron-l3-agent-snat -o qg-abce46ac-61 -j SNAT --to-source 10.10.0.240 --random-fully
+-A neutron-l3-agent-snat -m mark ! --mark 0x2/0xffff -m conntrack --ctstate DNAT -j SNAT --to-source 10.10.0.240 --random-fully
+-A neutron-postrouting-bottom -m comment --comment "Perform source NAT on outgoing traffic." -j neutron-l3-agent-snat
+COMMIT
+...
+# Completed on Mon Jan 24 14:28:27 2022
+# Generated by iptables-save v1.8.4 on Mon Jan 24 14:28:27 2022
+*mangle
+...
+-A neutron-l3-agent-mark -i qg-abce46ac-61 -j MARK --set-xmark 0x2/0xffff
+...
+```
+
+Also, traffic is send through the ``qg-`` interface with different source IP
+address. As there is no IP address assigned only to that specific Virtual
+Machine, router's external_gateway's address is used:
+
+```bash
+sudo ip netns exec qrouter-50711ca1-1e58-44dc-bd95-a87250d5b4ae tcpdump -i qg-abce46ac-61 -nvl
+tcpdump: listening on qg-abce46ac-61, link-type EN10MB (Ethernet), capture size 262144 bytes
+14:32:37.398487 IP (tos 0x0, ttl 63, id 41329, offset 0, flags [DF], proto ICMP (1), length 84)
+    10.10.0.240 > 8.8.8.8: ICMP echo request, id 6160, seq 0, length 64
+14:32:37.442202 IP (tos 0x80, ttl 112, id 0, offset 0, flags [none], proto ICMP (1), length 84)
+    8.8.8.8 > 10.10.0.240: ICMP echo reply, id 6160, seq 0, length 64
+```
+
+
+#### Bridge mappings
+
+I mentioned above that physical bridge, like "br-ex" needs to have physical
+interface plugged to be able to send traffic to the physical network.
+There can be more than one physical bridge in the node, if node has access to
+more than one physical network.
```

Comment:
I mentioned above that physical bridges, like "br-ex", need to have a physical
interface plugged to be able to send traffic to the corresponding physical network.
There can be more than one physical bridge in the node, if the node has access to
more than one physical network.